src/boot/efi/efi-string.c

```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                        *ret_field_width = field_width;
+                        return format;
+
+                case 'm':
+                        if (status <= ELEMENTSOF(warn_table) - 1) {
+                                *ret_str8 = warn_table[status];
+                                return format;
+                        }
+
+                        if (status >= EFI_ERROR_MASK &&
+                            status <= ((ELEMENTSOF(err_table) - 1) | EFI_ERROR_MASK)) {
+                                *ret_str8 = err_table[NOERR(status)];
+                                return format;
+                        }
+
+                        format_number(tmp,
+                                      /*min_digits=*/0,
+                                      NUMBER_HEX | NUMBER_HEX_PREFIX | NUMBER_LOWER,
+                                      status);
+                        *ret_str16 = tmp;
+                        return format;
+
+                case 'h':
+                        if (*(format + 1) == 'h')
+                                format++;
+
+                        /* Types smaller than int are promoted to int and would even elicit a warning when
+                         * fetched with va_arg. */
+                        field_length = sizeof(int);
+                        break;
+
+                case 'l':
+                        if (*(format + 1) == 'l') {
+                                format++;
+                                field_length = sizeof(long long);
+                        } else {
+                                long_arg = true;
+                                field_length = sizeof(long);
+                        }
+                        break;
+
+                case 'z':
+                        field_length = sizeof(size_t);
+                        break;
+
+                case 'd':
+                case 'i':
+                case 'u':
+                case 'x':
+                case 'X': {
+                        uint64_t v;
+                        NumberFormatFlags flags = NUMBER_DEFAULT;
+                        SET_FLAG(flags, NUMBER_HEX, IN_SET(*format, 'x', 'X'));
+                        SET_FLAG(flags, NUMBER_HEX_PREFIX, alternative_form);
+                        SET_FLAG(flags, NUMBER_SIGNED, IN_SET(*format, 'd', 'i'));
+                        SET_FLAG(flags, NUMBER_LOWER, IN_SET(*format, 'x'));
+
+                        switch (field_length) {
+                        case sizeof(int32_t):
+                                v = FLAGS_SET(flags, NUMBER_SIGNED) ? (uint64_t) va_arg(*ap, int32_t) :
+                                                                      va_arg(*ap, uint32_t);
+                                break;
+                        case sizeof(int64_t):
+                                v = FLAGS_SET(flags, NUMBER_SIGNED) ? va_arg(*ap, int64_t) :
+                                                                      va_arg(*ap, uint64_t);
```

Comment:
hmm, compilers don't warn that you return different types in both branches of the ternary op? weird...

Reply:
I think that only happens for pointers.